### ISSUE TITLE
jemalloc and ASAN / Valgrind don't get along

### ISSUE THREAD

**Issue by tli2** [cmu-db/noisepage#66]
ASAN and Valgrind currently do not catch any memory issues in tests.

As the title suggests linking in jemalloc results in some wonky issues where ASAN and Valgrind stops being able to detect leaks. The Internet suggests that either jemalloc doesn't expose the necessary interface or the dynamic linking interferes with instrumentation of the two tools. A quick search did not turn up any widely accepted explanation or fix.

See #56 for more information.

**Comment by tli2** [comment]
Temporarily fixed via #67. Should roll back once we figure out what's going on.

**Comment by mbutrovich** [comment]
There was some discussion today about linking in jemalloc only for benchmark executables since we don't check correctness in those anyway, but we decided that we're content (potentially) leaving performance on the table for now and can revisit jemalloc in the future when the system and benchmarks are a bit more robust.

**Comment by tli2** [comment]
We still want to eventually integrate with jemalloc as it makes us ~30% faster overall on transactions workload where we currently bottleneck on malloc.

**Comment by lmwnshn** [comment]
@mbutrovich You're right! : )

This ASAN enmity manifested as an oversized write when trying to emplace_back into a vector of strong_typedef'd uint16_t.

**Comment by lmwnshn** [comment]
I don't know what fixed it (ASAN side? jemalloc side?) but #1224 seems to run unittests with ASAN and jemalloc fine.

**Comment by lmwnshn** [comment]
2020-10-21: tested with ASAN and jemalloc in debug mode.

Still can detect stack buffer overflows.
```
int main(int argc, char *argv[]) {
  int a[8] = {0};
  for (int i = 0; i <= 8; ++i) {
    a[i] = 50;
  }
  (void) a;
  return 0;
}
```

Still can detect leaks.
```
int main(int argc, char *argv[]) {
  void *b = malloc(5000);
  (void)b;
  return 0;
}
```

Closing, looks fixed.
